wrappers/rust/icicle-core/src/polynomials/mod.rs

```diff
@@ -0,0 +1,707 @@
+                    }
+                }
+
+                pub fn sub_monomial_inplace(&mut self, monomial_coeff: &$field, monomial: u64) {
+                    unsafe {
+                        sub_monomial_inplace(self.handle, monomial_coeff, monomial);
+                    }
+                }
+
+                pub fn slice(&self, offset: u64, stride: u64, size: u64) -> Polynomial {
+                    unsafe {
+                        Polynomial {
+                            handle: slice(self.handle, offset, stride, size),
+                        }
+                    }
+                }
+
+                pub fn even(&self) -> Polynomial {
+                    unsafe {
+                        Polynomial {
+                            handle: even(self.handle),
+                        }
+                    }
+                }
+
+                pub fn odd(&self) -> Polynomial {
+                    unsafe {
+                        Polynomial {
+                            handle: odd(self.handle),
+                        }
+                    }
+                }
+
+                pub fn eval(&self, x: &$field) -> $field {
+                    unsafe { eval(self.handle, x) }
+                }
+
+                pub fn eval_on_domain<D: HostOrDeviceSlice<$field> + ?Sized, E: HostOrDeviceSlice<$field> + ?Sized>(
+                    &self,
+                    domain: &D,
+                    evals: &mut E,
+                ) {
+                    assert!(
+                        domain.len() <= evals.len(),
+                        "eval_on_domain(): eval size must not be smaller then domain"
+                    );
+                    unsafe {
+                        eval_on_domain(
+                            self.handle,
+                            domain.as_ptr(),
+                            domain.len() as u64,
+                            evals.as_mut_ptr(),
+                        );
+                    }
+                }
+
+                pub fn get_nof_coeffs(&self) -> u64 {
+                    unsafe {
+                        // returns total #coeffs. Not copying when null
+                        let nof_coeffs = copy_coeffs(self.handle, std::ptr::null_mut(), 0, 0);
+                        nof_coeffs as u64
+                    }
+                }
+
+                pub fn get_coeff(&self, idx: u64) -> $field {
+                    unsafe { get_coeff(self.handle, idx) }
+                }
+
+                pub fn copy_coeffs<S: HostOrDeviceSlice<$field> + ?Sized>(&self, start_idx: u64, coeffs: &mut S) {
+                    let coeffs_len = coeffs.len() as u64;
+                    let nof_coeffs = self.get_nof_coeffs();
+                    let end_idx = cmp::min(nof_coeffs, start_idx + coeffs_len - 1);
+
+                    unsafe {
+                        copy_coeffs(self.handle, coeffs.as_mut_ptr(), start_idx as i64, end_idx as i64);
```

Comment:
yeah it's a little clunky but I wanted u64 in rust API to avoid getting negative values. But at  the same type in C++ I wanted i64 since I can have an end_index be -1 (default value) which signals to copy [start:] without specifying the end value.